project/src/main/java/com/google/sps/data/ServletLibrary.java

```diff
@@ -201,4 +201,84 @@ public static Business retrieveBusinessInfo(DatastoreService datastore, String b
                         productIds,
                         tempVisionAnnotation);
   }
+
+  public static void addProductToLabels(DatastoreService datastore, String productId, List<String> labels) {
+    for (String label : labels) {
+      Filter filter = new FilterPredicate("label", FilterOperator.EQUAL, label.toLowerCase());
+      Query query = new Query("ProductLabel").setFilter(filter);
+      PreparedQuery pq = datastore.prepare(query);
+      Entity entity = pq.asSingleEntity();
+      // If the label doesnt already exist, create a new one and store. Otherwise,
+      // add the product id to the list contained in the existing entity.
+      if (entity == null) {
+        Entity productLabel = new Entity("ProductLabel");
+        productLabel.setProperty("label", label.toLowerCase());
+        List<String> productIds = new ArrayList<>();
+        productIds.add(productId);
+        productLabel.setProperty("productIds", productIds);
+        datastore.put(productLabel);
+      } else {
+        @SuppressWarnings("unchecked") // Documentation says to suppress warning this way
+          List<String> productIds = (ArrayList<String>) entity.getProperty("productIds"); 
```

Comment:
Maybe fix indentation here. I don't know if it's supposed to be like that because of the supress warning though.